addons/website_slides/controllers/main.py

```diff
@@ -1005,39 +1009,96 @@ def create_slide(self, *args, **post):
+                'id': slide.id,
+                'category_id': category_data['category']['id'],
+                'name': slide.name,
+                'slide_type': slide.slide_type,
+                'slug': slug(slide),
+                'is_preview': slide.is_preview,
+                'website_published': slide.website_published,
+                'action': action_id,
+                'edit_link': '/web#id=%s&action=%s&model=slide.slide&view_type=form' % (slide.id, action_id)
+            },
+        }
 
-        # handle exception during creation of slide and sent error notification to the client
+        res.update(category_data)
+
+        return res
+
+    def _check_channel_rights(self, channel_id):
+        # Handle exception during creation of slide and sent error notification to the client
         # otherwise client slide create dialog box continue processing even server fail to create a slide
         try:
-            channel = request.env['slide.channel'].browse(values['channel_id'])
-            can_upload = channel.can_upload
-            can_publish = channel.can_publish
+            channel = request.env['slide.channel'].browse(channel_id)
         except UserError as e:
             _logger.error(e)
             return {'error': e.args[0]}
         else:
-            if not can_upload:
+            if not channel.can_upload:
                 return {'error': _('You cannot upload on this channel.')}
+            else:
+                return channel
+
+    def _set_slide_create_values(self, post):
+        values = dict((fname, post[fname]) for fname in self._get_valid_slide_post_values() if post.get(fname))
 
         if post.get('duration'):
             # minutes to hours conversion
             values['completion_time'] = int(post['duration']) / 60
 
-        category_id = False
-        # handle creation of new categories on the fly
-        if post.get('category_id'):
-            category_id = post['category_id'][0]
-            if category_id == 0:
-                category = request.env['slide.slide'].create(self._get_new_slide_category_values(channel, post['category_id'][1]['name']))
-                values['sequence'] = category.sequence + 1
-            else:
-                category = request.env['slide.slide'].browse(category_id)
-                values.update({
-                    'sequence': request.env['slide.slide'].browse(post['category_id'][0]).sequence + 1
-                })
+        return values
 
-        # create slide itself
+    def _fetch_or_create_slide_category(self, channel, category_id=0, name=None):
+        category = request.env['slide.slide']
+        category_created = False
+        if category_id and not category_id == 0:
+            category = request.env['slide.slide'].browse(category_id)
+        elif category_id == 0:
+            category = request.env['slide.slide'].create(self._get_new_slide_category_values(channel, name))
+            category_created = True
```

Comment:
```
if category_id:
    ...
else:
    ...
```
as` category_id == 0` = `category_id == False`
no ?
unless category id can be negative ?